### ISSUE TITLE
[Bug] [keyof] Unexpected behavior of `keyof` of an interface with optional properties 

### ISSUE THREAD

**Issue by franklixuefei** [Microsoft/TypeScript#24679]
**TypeScript Version:**  2.9.1

**Search Terms:** 
keyof,optional
**Code**

```ts
export type Map<T = {}> = {
  readonly [P in Extract<keyof T, string>]: A<P>;
};

export interface A<TNameType extends string = string> {
    test: string;
}

export interface B {
    required: string;
    optional?: string;
}

export const Test: Map<B> = {
    required: 'required',
}

```

**Expected behavior:**

No errors should be reported, because an optional property should not be required as a result of `keyof`.

**Actual behavior:**

```
Type '{ required: string; }' is not assignable to type 'Map<B>'.
  Property 'optional' is missing in type '{ required: string; }'.
```

**It was working in 2.8.3, without `Extract<>`**

**Playground Link:**  
[Playground Link](https://www.typescriptlang.org/play/index.html#src=export%20type%20Map%3CT%20%3D%20%7B%7D%3E%20%3D%20%7B%0D%0A%20%20readonly%20%5BP%20in%20Extract%3Ckeyof%20T%2C%20string%3E%5D%3A%20A%3CP%3E%3B%0D%0A%7D%3B%0D%0A%0D%0Aexport%20interface%20A%3CTNameType%20extends%20string%20%3D%20string%3E%20%7B%0D%0A%20%20%20%20test%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20B%20%7B%0D%0A%20%20%20%20required%3A%20string%3B%0D%0A%20%20%20%20optional%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aexport%20const%20Test%3A%20Map%3CB%3E%20%3D%20%7B%0D%0A%20%20%20%20required%3A%20'required'%2C%0D%0A%7D%0D%0A)

**Related Issues:**

**Comment by DanielRosenwasser** [comment]
This is pretty unfortunate - our workaround to only include `string` keys was to use `Extract<keyof T, string>`, but mapped types that use that trick are no longer considered homomorphic mapped types, which means that we can't preserve modifiers on properties.

**Comment by DanielRosenwasser** [comment]
Is there a reason you only want to include `string` keys?

**Comment by jcalz** [comment]
I think if you use a layer of indirection you can get the desired homomorphicity back (homomorphicness? homomorphologicality?).  Specifically, make a new type parameter `K extends keyof T`, assign `Extract<keyof T, string>` to it, and then map over that.  Like this:

```ts
export type Map<T = {}, K extends keyof T = Extract<keyof T, string>> = {
  readonly [P in K]: A<P>;
};
```
or, equivalently
```ts
type OnlyStringKeys<T> = Pick<T, Extract<keyof T, string>>; 
export type Map<T = {}> = {
  readonly [P in keyof OnlyStringKeys<T>]: A<P>;
};
```

Then the following should work (note that the type of the `required` property had to be modified to type `{test: string}` as defined in `A<>` ; I assume OP wasn't meaning to make this error):

```ts
export const Test: Map<B> = {
  required: { test: 'required' },
};  // no error now
```

So at least there's a workaround?  Not sure if this is the way to do it going forward, or if there are other ideas.
